llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp

```diff
@@ -4056,6 +4056,58 @@ SDValue AMDGPUTargetLowering::performShlCombine(SDNode *N,
   SDLoc SL(N);
   SelectionDAG &DAG = DCI.DAG;
 
+  // When the shl64_reduce optimisation code is passed through vector
+  // legalization //some scalarising occurs. After ISD::AND was legalised, this
+  // resulted in the AND instructions no longer being elided, as mentioned
+  // below. The following code should make sure this takes place.
+  // ConstantSDNode *CVANDRHS = dyn_cast<ConstantSDNode>(RHS->getOperand(1));
```

Comment:
Does a similar problem for eliding AND also occur in performSrlCombine?  This should be tested in srl64_reduce.ll.

Also PR https://github.com/llvm/llvm-project/pull/144421 will add a similar 64-bit to 32-bit reduction for performSraCombine.

Reply:
PR https://github.com/llvm/llvm-project/pull/144421 has just been merged, so performSraCombine may also need to be updated to scalarise.